org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/WorkspaceDiagnosticsHandler.java

```diff
@@ -386,25 +395,55 @@ private static Range convertRange(IDocument document, IMarker marker) {
 				cStart = marker.getAttribute(IMavenConstants.MARKER_COLUMN_START, -1);
 				cEnd = marker.getAttribute(IMavenConstants.MARKER_COLUMN_END, -1);
 			} else {
-				int lineOffset = 0;
-				try {
-					lineOffset = document.getLineOffset(line);
-				} catch (BadLocationException unlikelyException) {
-					JavaLanguageServerPlugin.logException(unlikelyException.getMessage(), unlikelyException);
-					return new Range(new Position(line, 0), new Position(line, 0));
+				if (marker.getAttribute("id", -1) == IProblem.UndefinedType) {
+					try {
+						Range range = getAnnotationRange(document, marker);
+						if (range != null) {
+							return range;
+						}
+					} catch (BadLocationException e) {
+						JavaLanguageServerPlugin.logException(e.getMessage(), e);
+					}
+				} else {
+					int lineOffset = 0;
+					try {
+						lineOffset = document.getLineOffset(line);
+					} catch (BadLocationException unlikelyException) {
+						JavaLanguageServerPlugin.logException(unlikelyException.getMessage(), unlikelyException);
+						return new Range(new Position(line, 0), new Position(line, 0));
+					}
+					cEnd = marker.getAttribute(IMarker.CHAR_END, -1) - lineOffset;
+					cStart = marker.getAttribute(IMarker.CHAR_START, -1) - lineOffset;
 				}
-				cEnd = marker.getAttribute(IMarker.CHAR_END, -1) - lineOffset;
-				cStart = marker.getAttribute(IMarker.CHAR_START, -1) - lineOffset;
 			}
 		} catch (CoreException e) {
 			JavaLanguageServerPlugin.logException(e.getMessage(), e);
 		}
 		cStart = Math.max(0, cStart);
 		cEnd = Math.max(0, cEnd);
-
 		return new Range(new Position(line, cStart), new Position(line, cEnd));
 	}
 
+	private static Range getAnnotationRange(IDocument document, IMarker marker) throws BadLocationException, JavaModelException {
+		if (marker.getAttribute("id", -1) == IProblem.UndefinedType) {
+			int end = marker.getAttribute(IMarker.CHAR_END, -1);
+			int start = marker.getAttribute(IMarker.CHAR_START, -1);
+			while (start > 0) {
```

Comment:
If the while isn't needed here as from above, I would remove it in favour of an if.

Reply:
Fixed.